scipy/interpolate/src/_dierckxmodule.cc

```diff
@@ -710,6 +710,184 @@ py_find_interval(PyObject *self, PyObject *args)
+        "strides_c1 : ndarray, one-dimensional\n"
+        "    Pre-computed strides of the ``c1`` array.\n"
+        "    Note: These are *data* strides, not numpy-style byte strides.\n"
+        "    This array is equivalent to\n"
+        "    ``[stride // s1.dtype.itemsize for stride in s1.strides]``.\n"
+        "indices_k1d : ndarray, shape((k+1)**ndim, ndim)\n"
+        "    Pre-computed mapping between indices for iterating over a flattened\n"
+        "    array of shape ``[k[d] + 1) for d in range(ndim)`` and\n"
+        "    ndim-dimensional indices of the ``(k+1,)*ndim`` dimensional array.\n"
+        "    This is essentially a transposed version of\n"
+        "    ``np.unravel_index(np.arange((k+1)**ndim), (k+1,)*ndim)``.\n"
+        "\n"
+        "Returns\n"
+        "-------\n"
+        "out : ndarray, shape (npoints, num_c_tr)\n"
+        "    Output values of the b-spline at given ``xi`` points.\n"
+        "\n"
+        "Notes\n"
+        "-----\n"
+        "\n"
+        "This function is essentially equivalent to the following: given an\n"
+        "N-dimensional vector ``x = (x1, x2, ..., xN)``, iterate over the\n"
+        "dimensions, form linear combinations of products,\n"
+        "B(x1) * B(x2) * ... B(xN) of (k+1)**N b-splines which are non-zero\n"
+        "at ``x``.\n"
+        "\n"
+        "Since b-splines are localized, the sum has (k+1)**N non-zero elements.\n"
+        "\n"
+        "If ``i = (i1, i2, ..., iN)`` is a vector if intervals of the knot\n"
+        "vectors, ``t[d, id] <= xd < t[d, id+1]``, for ``d=1, 2, ..., N``, then\n"
+        "the core loop of this function is nothing but\n"
+        "\n"
+        "```\n"
+        "result = 0\n"
+        "iters = [range(i[d] - self.k[d], i[d] + 1) for d in range(ndim)]\n"
+        "for idx in itertools.product(*iters):\n"
+        "    term = self.c[idx] * np.prod([B(x[d], self.k[d], idx[d], self.t[d])\n"
+        "                                  for d in range(ndim)])\n"
+        "    result += term\n"
+        "```\n"
+        "\n"
+        "For efficiency reasons, we iterate over the flattened versions of the arrays.\n";
+/*
+def evaluate_ndbspline(const double[:, ::1] xi,
+                       const double[:, ::1] t,
+                       const npy_int64[::1] len_t,
+                       const npy_int64[::1] k,
+                       npy_int64[::1] nu,
+                       bint extrapolate,
+                       const double[::1] c1r,
+                       int num_c_tr,
+                       const npy_int64[::1] strides_c1,
+                       const npy_int64[:, ::] indices_k1d,
+*/
+static PyObject*
+py_evaluate_ndbspline(PyObject *self, PyObject *args)
+{
+    PyObject *py_xi=NULL;
+    PyObject *py_t=NULL, *py_c1r=NULL, *py_strides_c1=NULL, *py_indices_k1d=NULL;
+
+    PyObject *py_len_t=NULL, *py_k=NULL, *py_nu=NULL;
+    int num_c_tr;
+    int i_extrap;
+
+    if(!PyArg_ParseTuple(args, "OOOOOiOiOO",
```

Comment:
If you define py_t etc. as `PyArrayObject*` and use `O!` instead of plain `O` then you can get each with, say

```C
(!PyArg_ParseTuple(args,
                         ("O!O!i"),
                         &PyArray_Type, (PyObject **)&ap_a,
                         &PyArray_Type, (PyObject **)&ap_b,
                         &myint)
 ) { return NULL; }
``` 

Then the API will check for these inputs to be NumPy arrays so you don't need to cast them again below.

Reply:
"parse your arg tuple, oy!". Sorry could not help it  :-)

More seriously though, thanks for the suggestion. I guess I'd mark it for a follow-up improvement across the whole extension, not just these two functions.